### DISCUSSION TITLE
Writing multiple byte slices from socket using ffmpeg

### DISCUSSION THREAD

**Discussion by rhodee** [owncast/owncast#597] [Q&A]
I am really excited by the library and have a question and I hope someone can provide some insight. 

What is the appropriate `ffmpeg` invocation to write a chunk from a `MediaRecorder`?

My use case is as follows:

I have two cameras (laptop/external) displaying video in my browser and a `video` element and `MediaRecorder` for each. Every second data is sent via websocket to my server which has ffmpeg running. I am trying to adapt my `ffmpeg` invocation from this: https://owncast.online/docs/broadcasting/#using-with-ffmpeg.

For now, I just want to try one camera at a time. I assumed I should write to stdin using this command to see my stream:

```go
var (
	cmd = exec.Command("ffmpeg",
		"-re",
		"-framerate", "60",
		"-video_size", "640x480",
		"-i", "-",
		"-vcodec", "libx264",
		"-preset:v", "ultrafast",
		"-acodec", "aac",
		"-f", "flv",
		"rtmp://localhost:1935/live/STREAMING_KEY",
	)
)
```

and the handler code maps to this:

```go
// ws handler
		stdin, err := cmd.StdinPipe()
		if err != nil {
			log.Println("stream fail:", err)
		}

		started := false

		for {
			if msg, err = readMessage(); err != nil {
				break
			}

			if !started {
				if err := cmd.Start(); err != nil {
					log.Println("pipe fail:", err)
					break
				}
				started = true
			}

			if _, err := stdin.Write(msg); err != nil {
				log.Println("stream fail:", err)
			}
		}
```

It does not work as expected. The first concern was this message:

```sh
ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with Apple clang version 12.0.0 (clang-1200.0.32.27)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/4.3.1_4 --enable-shared --enable-pthreads --enable-version3 --enable-avresample --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librtmp --enable-libspeex --enable-libsoxr --enable-videotoolbox --disable-libjack --disable-indev=jack
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
Option framerate not found.
```

Then, every subsequent chunk results in this message being piped `stream fail: write |1: broken pipe`. 

Any guidance on using ffmpeg correctly (this is my first try) would be greatly appreciated.

**Comment by rhodee** [comment]
*Update*

It turns out this configuration does the trick 🍾 (latency is very high):

```go
	cmd = exec.Command(ffmpegPath,
		"-y",
		"-i", "-",
		"-c:v", "libx264",
		"-c:a", "aac",
		"-preset", "ultrafast",
		"-bufsize", "1000",
		"-f", "flv",
		"rtmp://localhost:1935/live/STREAMING_KEY",
	)
```

Can Owncast support two streams or is an intermediate step required to display both stream sources?

**Reply by gabek** [reply]
I'm glad you figured out what you were looking to do, because admittedly I didn't completely understand your use case and what you were working with :)

Owncast accepts a single RTMP source, if you wanted to mix multiple sources in that is best done on the broadcasting side.
